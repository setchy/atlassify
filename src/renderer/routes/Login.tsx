import { type FC, Fragment, useCallback, useContext, useState } from 'react';

import ButtonGroup from '@atlaskit/button/button-group';
import Button from '@atlaskit/button/new';
import Form, {
  ErrorMessage,
  Field,
  FormFooter,
  FormSection,
  HelperMessage,
} from '@atlaskit/form';
import LinkExternalIcon from '@atlaskit/icon/core/link-external';
import LogInIcon from '@atlaskit/icon/core/log-in';
import LockLockedIcon from '@atlaskit/icon/utility/lock-locked';
import { Box, Inline } from '@atlaskit/primitives';
import TextField from '@atlaskit/textfield';
import Tooltip from '@atlaskit/tooltip';

import { useNavigate } from 'react-router-dom';
import { logError } from '../../shared/logger';
import { Header } from '../components/primitives/Header';
import { AppContext } from '../context/App';
import type { Token, Username } from '../types';
import { checkIfCredentialsAreValid } from '../utils/api/client';
import type { LoginOptions } from '../utils/auth/types';
import {
  openAtlassianCreateToken,
  openAtlassianSecurityDocs,
} from '../utils/links';

interface IValues {
  username: Username;
  token: Token;
}

export const LoginRoute: FC = () => {
  const { login } = useContext(AppContext);
  const navigate = useNavigate();
  const [isValidToken, setIsValidToken] = useState<boolean>(true);

  const loginUser = useCallback(
    async (data: IValues) => {
      try {
        await checkIfCredentialsAreValid(data.username, data.token);

        await login(data as LoginOptions);
        navigate(-1);
      } catch (err) {
        logError('loginUser', 'failed to login with provided credentials', err);

        setIsValidToken(false);
      }
    },
    [login],
  );

  return (
    <Fragment>
      <Header>Login with Atlassian</Header>

      <Box paddingInline="space.400">
        <Form<IValues> onSubmit={loginUser}>
          {({ formProps, submitting }) => (
            <form {...formProps}>
              <FormSection>
                <Field
                  aria-required={true}
                  name="username"
                  label="Username"
                  defaultValue={''}
                  // FIXME: isRequired causes the renderer process on Windows devices to crash upon mouse enter
                  // isRequired
                  testId="login-username"
                >
                  {({ fieldProps }) => (
                    <Fragment>
                      <TextField autoComplete="off" {...fieldProps} />
                      <HelperMessage>
                        Your Atlassian username / email address
                      </HelperMessage>
                    </Fragment>
                  )}
                </Field>
                <Field
                  aria-required={true}
                  name="token"
                  label="API Token"
                  defaultValue={''}
                  // FIXME: isRequired causes the renderer process on Windows devices to crash upon mouse enter
                  // isRequired
                  testId="login-token"
                >
                  {({ fieldProps }) => (
                    <Fragment>
                      <TextField type="password" {...fieldProps} />
                      <HelperMessage>
                        <Inline alignBlock="center" space="space.050">
                          <Tooltip content="Create an API Token">
                            <Button
                              appearance="discovery"
                              spacing="compact"
                              iconBefore={LockLockedIcon}
                              onClick={() => openAtlassianCreateToken()}
                              testId="login-create-token"
                            >
                              Create an API Token
                            </Button>
                          </Tooltip>
                          <Box>for your account and paste above</Box>
                        </Inline>
                      </HelperMessage>
                    </Fragment>
                  )}
                </Field>
              </FormSection>

              <Box paddingBlock="space.050">
                {!isValidToken && (
                  <ErrorMessage>
                    Oops! The username + token combination provided are not
                    valid. Please try again.
                  </ErrorMessage>
                )}
                <Inline alignBlock="center" spread="space-between">
                  <Box paddingBlockStart="space.300">
                    <Tooltip
                      content="See Atlassian documentation"
                      position="top"
                    >
                      <Button
                        appearance="subtle"
                        iconBefore={LinkExternalIcon}
                        onClick={() => openAtlassianSecurityDocs()}
                        testId="login-docs"
                      >
                        Docs
                      </Button>
                    </Tooltip>
                  </Box>
                  <FormFooter>
                    <ButtonGroup label="Form submit options">
                      <Button
                        appearance="subtle"
                        onClick={() => navigate(-1)}
                        testId="login-cancel"
                      >
                        Cancel
                      </Button>
                      <Button
                        type="submit"
                        iconBefore={LogInIcon}
                        appearance="primary"
                        isLoading={submitting}
                        testId="login-submit"
                      >
                        Login
                      </Button>
                    </ButtonGroup>
                  </FormFooter>
                </Inline>
              </Box>
            </form>
          )}
        </Form>
      </Box>
    </Fragment>
  );
};
