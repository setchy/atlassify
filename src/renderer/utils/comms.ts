import { defaultSettings } from '../context/defaults';
import { type Link, OpenPreference } from '../types';
import { loadState } from './storage';

export function openExternalLink(url: Link): void {
  // Load the state from local storage to avoid having to pass settings as a parameter
  const { settings } = loadState();
  const openPreference = settings
    ? settings.openLinks
    : defaultSettings.openLinks;

  if (url.toLowerCase().startsWith('https://')) {
    window.atlassify.openExternalLink(
      url,
      openPreference === OpenPreference.FOREGROUND,
    );
  }
}

export async function getAppVersion(): Promise<string> {
  return await window.atlassify.app.version();
}

export async function encryptValue(value: string): Promise<string> {
  return await window.atlassify.encryptValue(value);
}

export async function decryptValue(value: string): Promise<string> {
  return await window.atlassify.decryptValue(value);
}

export function quitApp(): void {
  window.atlassify.app.quit();
}

export function showWindow(): void {
  window.atlassify.app.show();
}

export function hideWindow(): void {
  window.atlassify.app.hide();
}

export function setAutoLaunch(value: boolean): void {
  window.atlassify.setAutoLaunch(value);
}

export function setUseAlternateIdleIcon(value: boolean): void {
  window.atlassify.tray.useAlternateIdleIcon(value);
}

export function setUseUnreadActiveIcon(value: boolean): void {
  window.atlassify.tray.useUnreadActiveIcon(value);
}

export function setKeyboardShortcut(keyboardShortcut: boolean): void {
  window.atlassify.setKeyboardShortcut(keyboardShortcut);
}

export function updateTrayIcon(notificationsLength = 0): void {
  window.atlassify.tray.updateIcon(notificationsLength);
}

export function updateTrayTitle(title = ''): void {
  window.atlassify.tray.updateTitle(title);
}
