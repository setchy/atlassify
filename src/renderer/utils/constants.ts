export const Constants = {
  REPO_SLUG: 'setchy/atlassify',

  STORAGE_KEY: 'atlassify-storage',

  ALL_READ_EMOJIS: ['🎉', '🎊', '🥳', '👏', '🙌', '😎', '🏖️', '🚀', '✨', '🏆'],

  FETCH_NOTIFICATIONS_INTERVAL: 60000,

  REFRESH_ACCOUNTS_INTERVAL: 3600000,

  MAX_NOTIFICATIONS_PER_ACCOUNT: 999,
};
