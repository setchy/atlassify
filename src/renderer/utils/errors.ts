import type { AtlassifyError, ErrorType } from '../types';

export const Errors: Record<ErrorType, AtlassifyError> = {
  BAD_CREDENTIALS: {
    title: 'Bad Credentials',
    descriptions: [
      'Your credentials are either invalid or expired.',
      'Please try removing your account and authenticating again.',
    ],
    emojis: ['ðŸ”“'],
  },
  BAD_REQUEST: {
    title: 'Bad Request',
    descriptions: ['Something went wrong making the API request.'],
    emojis: ['ðŸ˜³'],
  },
  NETWORK: {
    title: 'Network Error',
    descriptions: [
      'Unable to connect to Atlassian Cloud.',
      'Please check your network connection, including whether you require a VPN, and try again.',
    ],
    emojis: ['ðŸ›œ'],
  },
  UNKNOWN: {
    title: 'Oops! Something went wrong',
    descriptions: ['Please try again later.'],
    emojis: ['ðŸ¤”', 'ðŸ¥²', 'ðŸ« ', 'ðŸ™ƒ', 'ðŸ™ˆ'],
  },
};
