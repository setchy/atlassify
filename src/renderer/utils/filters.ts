import EmojiFlagsIcon from '@atlaskit/icon/glyph/emoji/flags';
import WatchIcon from '@atlaskit/icon/glyph/watch';

import type {
  AccountNotifications,
  Category,
  FilterDetails,
  ProductName,
  ReadState,
  SettingsState,
} from '../types';

export const CATEGORIES: Record<Category, FilterDetails> = {
  direct: {
    name: 'direct',
    description: 'Direct notification',
    icon: EmojiFlagsIcon,
  },
  watching: {
    name: 'watching',
    description: 'Watching notification',
    icon: WatchIcon,
  },
};

export function getCategoryDetails(category: Category): FilterDetails {
  return CATEGORIES[category];
}

export const READ_STATES: Record<ReadState, FilterDetails> = {
  unread: {
    name: 'unread',
    description: 'Unread notification',
  },
  read: {
    name: 'read',
    description: 'Read notification',
  },
};

export function getReadStateDetails(readState: ReadState): FilterDetails {
  return READ_STATES[readState];
}

export function getReadStateFilterCount(
  notifications: AccountNotifications[],
  readState: ReadState,
) {
  return notifications.reduce(
    (memo, acc) =>
      memo + acc.notifications.filter((n) => n.readState === readState).length,
    0,
  );
}

export function getCategoryFilterCount(
  notifications: AccountNotifications[],
  category: Category,
) {
  return notifications.reduce(
    (memo, acc) =>
      memo + acc.notifications.filter((n) => n.category === category).length,
    0,
  );
}

export function getProductFilterCount(
  notifications: AccountNotifications[],
  product: ProductName,
) {
  return notifications.reduce(
    (memo, acc) =>
      memo + acc.notifications.filter((n) => n.product.name === product).length,
    0,
  );
}

export function hasFiltersSet(settings: SettingsState): boolean {
  return (
    settings.filterCategories.length > 0 ||
    settings.filterReadStates.length > 0 ||
    settings.filterProducts.length > 0
  );
}
