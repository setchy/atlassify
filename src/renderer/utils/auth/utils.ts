import log from 'electron-log';
import type { Account, AuthState, Token, Username } from '../../types';
import { getAuthenticatedUser } from '../api/client';
import { encryptValue } from '../comms';

export async function addAccount(
  auth: AuthState,
  username: Username,
  token: Token,
): Promise<AuthState> {
  const encryptedToken = await encryptValue(token);

  let newAccount = {
    username: username,
    token: encryptedToken,
  } as Account;

  // Refresh user data
  newAccount = await refreshAccount(newAccount);

  return {
    accounts: [...auth.accounts, newAccount],
  };
}

export function removeAccount(auth: AuthState, account: Account): AuthState {
  const updatedAccounts = auth.accounts.filter((a) => a.id !== account.id);

  return {
    accounts: updatedAccounts,
  };
}

export async function refreshAccount(account: Account): Promise<Account> {
  try {
    const res = await getAuthenticatedUser(account);

    account.id = res.data.data.me.user.accountId;
    account.name = res.data.data.me.user.name;
    account.avatar = res.data.data.me.user.picture;
  } catch (error) {
    log.error('Failed to refresh account', error);
  }

  return account;
}

export function hasAccounts(auth: AuthState) {
  return auth.accounts.length > 0;
}
