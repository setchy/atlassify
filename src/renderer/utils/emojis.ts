import twemoji, { type TwemojiOptions } from "@discordapp/twemoji";

import { Constants } from "./constants";
import { Errors } from "./errors";

const EMOJI_FORMAT = "svg";

const ALL_EMOJIS = [
	...Constants.ALL_READ_EMOJIS,
	...Errors.BAD_CREDENTIALS.emojis,
	...Errors.BAD_REQUEST.emojis,
	...Errors.NETWORK.emojis,
	...Errors.UNKNOWN.emojis,
];

export const ALL_EMOJI_SVG_FILENAMES = ALL_EMOJIS.map(async (emoji) => {
	const imgHtml = convertTextToEmojiImgHtml(emoji);
	return extractSvgFilename(await imgHtml);
});

export async function convertTextToEmojiImgHtml(text: string): Promise<string> {
	// return text;

	const directory = await window.atlassify.twemojiDirectory();

	console.log("ADAM DIR", directory);

	return twemoji.parse(text, {
		folder: EMOJI_FORMAT,
		callback: (icon: string, _options: TwemojiOptions) => {
			return `${directory}/${icon}.${EMOJI_FORMAT}`;
		},
	});
}

function extractSvgFilename(imgHtml: string): string {
	const srcMatch = /src="(.*)"/.exec(imgHtml);
	const src = srcMatch ? srcMatch[1] : "";
	const filename = src.split("/").pop(); // Get the last part after splitting by "/"
	return filename;
}
