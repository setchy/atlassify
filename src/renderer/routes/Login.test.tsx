import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import * as comms from '../utils/comms';
import { LoginRoute } from './Login';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('renderer/routes/Login.tsx', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render itself & its children', () => {
    const tree = render(<LoginRoute />);

    expect(tree).toMatchSnapshot();
  });

  describe('login web pages', () => {
    it('should open create new token page', async () => {
      const openExternalLinkMock = jest
        .spyOn(comms, 'openExternalLink')
        .mockImplementation();

      await act(async () => {
        render(<LoginRoute />);
      });

      await userEvent.click(screen.getByTestId('login-create-token'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
    });
    it('should open login docs', async () => {
      const openExternalLinkMock = jest
        .spyOn(comms, 'openExternalLink')
        .mockImplementation();

      await act(async () => {
        render(<LoginRoute />);
      });

      await userEvent.click(screen.getByTestId('login-docs'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
    });
  });

  it('should navigate back to landing page on cancel', async () => {
    await act(async () => {
      render(<LoginRoute />);
    });

    await userEvent.click(screen.getByTestId('login-cancel'));

    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });
});
