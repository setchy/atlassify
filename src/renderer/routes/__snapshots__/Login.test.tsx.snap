// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer/routes/Login.tsx should render itself & its children 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-18bes35"
      >
        <div
          class="css-y272c0"
        >
          <div
            role="presentation"
          >
            <button
              class="css-1t6wxle"
              data-testid="header-nav-back"
              type="button"
            >
              <span
                class="css-1spmf3f"
              >
                <span
                  aria-hidden="true"
                  class="css-snhnyn"
                  data-vc="icon-undefined"
                  style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                >
                  <svg
                    height="24"
                    role="presentation"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M12.207 5.793a1 1 0 0 1 0 1.414L8.414 11H18a1 1 0 1 1 0 2H8.414l3.793 3.793a1 1 0 1 1-1.415 1.414l-5.499-5.5A1 1 0 0 1 5 12.028v-.057a1 1 0 0 1 .293-.678l5.499-5.5a1 1 0 0 1 1.415 0"
                      fill="currentcolor"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="css-b5o75w"
                >
                  Go Back
                </span>
              </span>
            </button>
          </div>
          <h3
            class="css-rex7s6"
          >
            Login with Atlassian
          </h3>
        </div>
      </div>
      <div
        class="css-1rd5nai"
      >
        <form>
          <div
            class="css-2pqcrg"
          >
            <div
              class="css-wwet8f"
              data-testid="login-username"
            >
              <label
                class="css-g6a23j"
                data-testid="login-username--label"
                for="username-uid3"
                id="username-uid3-label"
              >
                Username
                <span
                  aria-hidden="true"
                  class="css-1qy5j0p"
                  title="required"
                >
                  *
                </span>
              </label>
              <div
                class="css-17nf42q"
                data-ds--text-field--container="true"
                role="presentation"
              >
                <input
                  aria-describedby="username-uid3-valid username-uid3-helper"
                  aria-labelledby="username-uid3-label"
                  autocomplete="off"
                  class="css-1kxou5n"
                  data-ds--text-field--input="true"
                  id="username-uid3"
                  name="username"
                  required=""
                  value=""
                />
              </div>
              <div
                aria-live="polite"
                class="css-5cg24a"
                id="username-uid3-helper"
              >
                Your Atlassian username / email address
              </div>
            </div>
            <div
              class="css-wwet8f"
              data-testid="login-token"
            >
              <label
                class="css-g6a23j"
                data-testid="login-token--label"
                for="token-uid4"
                id="token-uid4-label"
              >
                API Token
                <span
                  aria-hidden="true"
                  class="css-1qy5j0p"
                  title="required"
                >
                  *
                </span>
              </label>
              <div
                class="css-17nf42q"
                data-ds--text-field--container="true"
                role="presentation"
              >
                <input
                  aria-describedby="token-uid4-valid token-uid4-helper"
                  aria-labelledby="token-uid4-label"
                  class="css-1kxou5n"
                  data-ds--text-field--input="true"
                  id="token-uid4"
                  name="token"
                  required=""
                  type="password"
                  value=""
                />
              </div>
              <div
                aria-live="polite"
                class="css-5cg24a"
                id="token-uid4-helper"
              >
                <span>
                  <div
                    class="css-bsfd3j"
                  >
                    <div
                      role="presentation"
                    >
                      <button
                        class="css-9v2r3w"
                        data-testid="login-create-token"
                        type="button"
                      >
                        <span
                          class="css-12rb8n9"
                        >
                          <span
                            aria-hidden="true"
                            class="css-snhnyn"
                            data-vc="icon-undefined"
                            style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                          >
                            <svg
                              height="24"
                              role="presentation"
                              viewBox="0 0 24 24"
                              width="24"
                            >
                              <g
                                fill="currentcolor"
                              >
                                <path
                                  d="M16 11V9h-2V7.002A2 2 0 0 0 12 5c-1.102 0-2 .898-2 2.002V9H8v2H7v8h10v-8zm-2 0h-4V9h4zM8 9V7.002A4.004 4.004 0 0 1 12 3a4 4 0 0 1 4 4.002V9h.994A2.01 2.01 0 0 1 19 11.009v7.982c0 1.11-.897 2.009-2.006 2.009H7.006A2.01 2.01 0 0 1 5 18.991V11.01C5 9.899 5.897 9 7.006 9zm0 0h2v2H8zm6 0h2v2h-2z"
                                />
                                <circle
                                  cx="12"
                                  cy="15"
                                  r="2"
                                />
                              </g>
                            </svg>
                          </span>
                        </span>
                        <span
                          class="css-1nqby2s"
                        >
                          Create an API Token
                        </span>
                      </button>
                    </div>
                    <div
                      class="css-k5cxwk"
                    >
                      for your account and paste above
                    </div>
                  </div>
                </span>
              </div>
            </div>
          </div>
          <div
            class="css-1rw5qv2"
          >
            <div
              class="css-kxmhz3"
            >
              <div
                class="css-1gbz6oi"
              >
                <div
                  role="presentation"
                >
                  <button
                    class="css-vgqjed"
                    data-testid="login-docs"
                    type="button"
                  >
                    <span
                      class="css-12rb8n9"
                    >
                      <span
                        aria-hidden="true"
                        class="css-snhnyn"
                        data-vc="icon-undefined"
                        style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                      >
                        <svg
                          height="24"
                          role="presentation"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <g
                            fill="currentcolor"
                          >
                            <path
                              d="M5 19.01V19zM5 4.99V5zM19 19v-6h2v6a2 2 0 0 1-1.99 2H5a2 2 0 0 1-2-1.99V4.99C3 3.89 3.9 3 5 3h6v2H5v14zM5 4.99V5zM11 5H5v14h14v-6h2v6a2 2 0 0 1-1.99 2H5a2 2 0 0 1-2-1.99V4.99C3 3.89 3.9 3 5 3h6zm8 0v3a1 1 0 0 0 2 0V4a1 1 0 0 0-1-1h-4a1 1 0 0 0 0 2z"
                            />
                            <path
                              d="m12.707 12.707 8-8a1 1 0 1 0-1.414-1.414l-8 8a1 1 0 0 0 1.414 1.414"
                            />
                          </g>
                        </svg>
                      </span>
                    </span>
                    <span
                      class="css-1nqby2s"
                    >
                      Docs
                    </span>
                  </button>
                </div>
              </div>
              <footer
                class="css-kwugxo"
              >
                <div
                  aria-label="Form submit options"
                  class="css-1oy5iav"
                  role="group"
                >
                  <button
                    class="css-eu6vk3"
                    data-testid="login-cancel"
                    type="button"
                  >
                    <span
                      class="css-1nqby2s"
                    >
                      Cancel
                    </span>
                  </button>
                  <button
                    class="css-1vvzp1k"
                    data-testid="login-submit"
                    type="submit"
                  >
                    <span
                      class="css-12rb8n9"
                    >
                      <span
                        aria-hidden="true"
                        class="css-snhnyn"
                        data-vc="icon-undefined"
                        style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                      >
                        <svg
                          height="24"
                          role="presentation"
                          viewBox="0 0 24 24"
                          width="24"
                        >
                          <g
                            fill="currentcolor"
                            fill-rule="evenodd"
                          >
                            <path
                              d="M15.503 3H12v2h7v14h-7v2h7.006c1.1 0 1.994-.893 1.994-1.995V4.995A1.993 1.993 0 0 0 19.006 3z"
                            />
                            <path
                              d="M3.977 11A.99.99 0 0 0 3 12.001c0 .551.437.999.977.999h11.047a.99.99 0 0 0 .976-.999.99.99 0 0 0-.976-1.001z"
                            />
                            <path
                              d="M12.303 8.305a1.053 1.053 0 0 0 0 1.478L14.5 12l-2.197 2.217a1.05 1.05 0 0 0 0 1.476c.404.409 1.06.409 1.465 0l2.93-2.955a1.055 1.055 0 0 0 0-1.477l-2.93-2.956a1.034 1.034 0 0 0-1.465 0"
                            />
                          </g>
                        </svg>
                      </span>
                    </span>
                    <span
                      class="css-1nqby2s"
                    >
                      Login
                    </span>
                  </button>
                </div>
              </footer>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-18bes35"
    >
      <div
        class="css-y272c0"
      >
        <div
          role="presentation"
        >
          <button
            class="css-1t6wxle"
            data-testid="header-nav-back"
            type="button"
          >
            <span
              class="css-1spmf3f"
            >
              <span
                aria-hidden="true"
                class="css-snhnyn"
                data-vc="icon-undefined"
                style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
              >
                <svg
                  height="24"
                  role="presentation"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    clip-rule="evenodd"
                    d="M12.207 5.793a1 1 0 0 1 0 1.414L8.414 11H18a1 1 0 1 1 0 2H8.414l3.793 3.793a1 1 0 1 1-1.415 1.414l-5.499-5.5A1 1 0 0 1 5 12.028v-.057a1 1 0 0 1 .293-.678l5.499-5.5a1 1 0 0 1 1.415 0"
                    fill="currentcolor"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="css-b5o75w"
              >
                Go Back
              </span>
            </span>
          </button>
        </div>
        <h3
          class="css-rex7s6"
        >
          Login with Atlassian
        </h3>
      </div>
    </div>
    <div
      class="css-1rd5nai"
    >
      <form>
        <div
          class="css-2pqcrg"
        >
          <div
            class="css-wwet8f"
            data-testid="login-username"
          >
            <label
              class="css-g6a23j"
              data-testid="login-username--label"
              for="username-uid3"
              id="username-uid3-label"
            >
              Username
              <span
                aria-hidden="true"
                class="css-1qy5j0p"
                title="required"
              >
                *
              </span>
            </label>
            <div
              class="css-17nf42q"
              data-ds--text-field--container="true"
              role="presentation"
            >
              <input
                aria-describedby="username-uid3-valid username-uid3-helper"
                aria-labelledby="username-uid3-label"
                autocomplete="off"
                class="css-1kxou5n"
                data-ds--text-field--input="true"
                id="username-uid3"
                name="username"
                required=""
                value=""
              />
            </div>
            <div
              aria-live="polite"
              class="css-5cg24a"
              id="username-uid3-helper"
            >
              Your Atlassian username / email address
            </div>
          </div>
          <div
            class="css-wwet8f"
            data-testid="login-token"
          >
            <label
              class="css-g6a23j"
              data-testid="login-token--label"
              for="token-uid4"
              id="token-uid4-label"
            >
              API Token
              <span
                aria-hidden="true"
                class="css-1qy5j0p"
                title="required"
              >
                *
              </span>
            </label>
            <div
              class="css-17nf42q"
              data-ds--text-field--container="true"
              role="presentation"
            >
              <input
                aria-describedby="token-uid4-valid token-uid4-helper"
                aria-labelledby="token-uid4-label"
                class="css-1kxou5n"
                data-ds--text-field--input="true"
                id="token-uid4"
                name="token"
                required=""
                type="password"
                value=""
              />
            </div>
            <div
              aria-live="polite"
              class="css-5cg24a"
              id="token-uid4-helper"
            >
              <span>
                <div
                  class="css-bsfd3j"
                >
                  <div
                    role="presentation"
                  >
                    <button
                      class="css-9v2r3w"
                      data-testid="login-create-token"
                      type="button"
                    >
                      <span
                        class="css-12rb8n9"
                      >
                        <span
                          aria-hidden="true"
                          class="css-snhnyn"
                          data-vc="icon-undefined"
                          style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                        >
                          <svg
                            height="24"
                            role="presentation"
                            viewBox="0 0 24 24"
                            width="24"
                          >
                            <g
                              fill="currentcolor"
                            >
                              <path
                                d="M16 11V9h-2V7.002A2 2 0 0 0 12 5c-1.102 0-2 .898-2 2.002V9H8v2H7v8h10v-8zm-2 0h-4V9h4zM8 9V7.002A4.004 4.004 0 0 1 12 3a4 4 0 0 1 4 4.002V9h.994A2.01 2.01 0 0 1 19 11.009v7.982c0 1.11-.897 2.009-2.006 2.009H7.006A2.01 2.01 0 0 1 5 18.991V11.01C5 9.899 5.897 9 7.006 9zm0 0h2v2H8zm6 0h2v2h-2z"
                              />
                              <circle
                                cx="12"
                                cy="15"
                                r="2"
                              />
                            </g>
                          </svg>
                        </span>
                      </span>
                      <span
                        class="css-1nqby2s"
                      >
                        Create an API Token
                      </span>
                    </button>
                  </div>
                  <div
                    class="css-k5cxwk"
                  >
                    for your account and paste above
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
        <div
          class="css-1rw5qv2"
        >
          <div
            class="css-kxmhz3"
          >
            <div
              class="css-1gbz6oi"
            >
              <div
                role="presentation"
              >
                <button
                  class="css-vgqjed"
                  data-testid="login-docs"
                  type="button"
                >
                  <span
                    class="css-12rb8n9"
                  >
                    <span
                      aria-hidden="true"
                      class="css-snhnyn"
                      data-vc="icon-undefined"
                      style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                    >
                      <svg
                        height="24"
                        role="presentation"
                        viewBox="0 0 24 24"
                        width="24"
                      >
                        <g
                          fill="currentcolor"
                        >
                          <path
                            d="M5 19.01V19zM5 4.99V5zM19 19v-6h2v6a2 2 0 0 1-1.99 2H5a2 2 0 0 1-2-1.99V4.99C3 3.89 3.9 3 5 3h6v2H5v14zM5 4.99V5zM11 5H5v14h14v-6h2v6a2 2 0 0 1-1.99 2H5a2 2 0 0 1-2-1.99V4.99C3 3.89 3.9 3 5 3h6zm8 0v3a1 1 0 0 0 2 0V4a1 1 0 0 0-1-1h-4a1 1 0 0 0 0 2z"
                          />
                          <path
                            d="m12.707 12.707 8-8a1 1 0 1 0-1.414-1.414l-8 8a1 1 0 0 0 1.414 1.414"
                          />
                        </g>
                      </svg>
                    </span>
                  </span>
                  <span
                    class="css-1nqby2s"
                  >
                    Docs
                  </span>
                </button>
              </div>
            </div>
            <footer
              class="css-kwugxo"
            >
              <div
                aria-label="Form submit options"
                class="css-1oy5iav"
                role="group"
              >
                <button
                  class="css-eu6vk3"
                  data-testid="login-cancel"
                  type="button"
                >
                  <span
                    class="css-1nqby2s"
                  >
                    Cancel
                  </span>
                </button>
                <button
                  class="css-1vvzp1k"
                  data-testid="login-submit"
                  type="submit"
                >
                  <span
                    class="css-12rb8n9"
                  >
                    <span
                      aria-hidden="true"
                      class="css-snhnyn"
                      data-vc="icon-undefined"
                      style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
                    >
                      <svg
                        height="24"
                        role="presentation"
                        viewBox="0 0 24 24"
                        width="24"
                      >
                        <g
                          fill="currentcolor"
                          fill-rule="evenodd"
                        >
                          <path
                            d="M15.503 3H12v2h7v14h-7v2h7.006c1.1 0 1.994-.893 1.994-1.995V4.995A1.993 1.993 0 0 0 19.006 3z"
                          />
                          <path
                            d="M3.977 11A.99.99 0 0 0 3 12.001c0 .551.437.999.977.999h11.047a.99.99 0 0 0 .976-.999.99.99 0 0 0-.976-1.001z"
                          />
                          <path
                            d="M12.303 8.305a1.053 1.053 0 0 0 0 1.478L14.5 12l-2.197 2.217a1.05 1.05 0 0 0 0 1.476c.404.409 1.06.409 1.465 0l2.93-2.955a1.055 1.055 0 0 0 0-1.477l-2.93-2.956a1.034 1.034 0 0 0-1.465 0"
                          />
                        </g>
                      </svg>
                    </span>
                  </span>
                  <span
                    class="css-1nqby2s"
                  >
                    Login
                  </span>
                </button>
              </div>
            </footer>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
